extraction_history/Extraction_tool.re

```diff
@@ -0,0 +1,217 @@
+// this is the tool functions
+
+open Extraction_declear;
+
+// FIXME: A temporary solution to map Some(Hole) into None
+let hole_to_none = (~uht: option(UHTyp.t)): option(UHTyp.t) =>
+  switch (uht) {
+  | None => None
+  | Some(t) =>
+    switch (t) {
+    | OpSeq(_, S(Hole, E)) => None
+    | _ => uht
+    }
+  };
+
+let rec option_string_concat =
+        (~strs: list(option(string))): option(string) =>
+  switch (strs) {
+  | [] => Some("")
+  | [a, ...rest] =>
+    switch (a, option_string_concat(~strs=rest)) {
+    | (Some(s1), Some(s2)) => Some(s1 ++ s2)
+    | _ => None
+    }
+  };
+
+let rec option_string_concat_ignoreNone =
+        (~strs: list(option(string))): option(string) =>
+  switch (strs) {
+  | [] => Some("")
+  | [a, ...rest] =>
+    switch (a, option_string_concat_ignoreNone(~strs=rest)) {
+    | (Some(s1), Some(s2)) => Some(s1 ++ s2)
+    | (Some(s1), None) => Some(s1)
+    | (None, Some(s2)) => Some(s2)
+    | _ => None
+    }
+  };
+
+//insert indent_space just like previous work
+let rec indent_space = (~level: int): string =>
+  if (level > 0) {
+    "  " ++ indent_space(~level=level - 1);
```

Comment:
probably easier to just run ocaml's autoformatter on the output than try to manage layout/indentation manually